kubetest/gke.go

```diff
@@ -608,12 +617,136 @@ func (g *gkeDeployer) cleanupNetworkFirewalls() (int, error) {
+		"--format=value(name)")) != nil {
+		log.Printf("Couldn't describe nat '%s', assuming it doesn't exist and creating it", nat)
+		if err := control.FinishRunning(exec.Command("gcloud", "compute", "routers", "nats", "create", nat,
+			"--project="+g.project,
+			"--router="+nat,
+			"--router-region="+region,
+			"--auto-allocate-nat-external-ips",
+			"--nat-all-subnet-ip-ranges",
+			"--min-ports-per-vm="+strconv.Itoa(g.natMinPortsPerVm),
+			"--enable-logging")); err != nil {
+			return fmt.Errorf("error creating NAT config: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (g *gkeDeployer) getRegion(region, zone string) (string, error) {
+	if region != "" {
+		return region, nil
+	}
+	result, err := exec.Command("gcloud", "compute", "zones", "list",
+		"--filter=name="+zone,
+		"--format=value(region)",
+		"--project="+g.project).Output()
+	if err != nil {
+		return "", fmt.Errorf("error resolving region of %s zone: %v", zone, err)
+	}
+	return strings.TrimSuffix(string(result), "\n"), nil
+}
+
+func (g *gkeDeployer) getNatName() (string, error) {
+	if err := g.getInstanceGroups(); err != nil {
+		return "", err
+	}
+	// Create unique NAT router name based on hash
+	// from the lexically first node pool instead.
+	return "nat-router-" + g.instanceGroups[0].uniq, nil
+}
+
+func (g *gkeDeployer) cleanupNat(nat string) error {
+	if !g.nat {
+		return nil
+	}
+	region, err := g.getRegion(g.region, g.zone)
+	if err != nil {
+		return fmt.Errorf("error finding region for NAT router: %v", err)
+	}
+
+	if control.NoOutput(exec.Command("gcloud", "compute", "routers", "nats", "describe", nat,
+		"--project="+g.project,
+		"--router="+nat,
+		"--router-region="+region,
+		"--format=value(name)")) == nil {
+		log.Printf("Found NAT configuration '%s', deleting", nat)
+		err = control.FinishRunning(exec.Command("gcloud", "compute", "routers", "nats", "delete", "-q", nat,
+			"--project="+g.project,
+			"--router="+nat,
+			"--router-region="+region))
+		if err != nil {
+			return fmt.Errorf("error deleting NAT config: %v", err)
+		}
+	} else {
+		log.Printf("Found no NAT configuration '%s', assuming resources are clean", nat)
+	}
+
+	if control.NoOutput(exec.Command("gcloud", "compute", "routers", "describe", nat,
+		"--project="+g.project,
+		"--region="+region,
+		"--format=value(name)")) == nil {
+		log.Printf("Found NAT router '%s', deleting", nat)
+		err = control.FinishRunning(exec.Command("gcloud", "compute", "routers", "delete", "-q", nat,
+			"--project="+g.project,
+			"--region="+region))
+		if err != nil {
+			return fmt.Errorf("error deleting NAT router: %v", err)
+		}
+	} else {
+		log.Printf("Found no NAT router '%s', assuming resources are clean", nat)
+	}
+
+	return nil
+}
+
 func (g *gkeDeployer) Down() error {
 	firewall, err := g.getClusterFirewall()
 	if err != nil {
 		// This is expected if the cluster doesn't exist.
 		return nil
 	}
+	nat, err := g.getNatName()
+	if err != nil {
+		// This is expected if the cluster doesn't exist.
```

Comment:
I don't like this, getNatName returns error whenever getInstanceGroups returns error and the latter doesn't return error only when cluster doesn't exist.

Reply:
Changed

Reply:
Now this if clause is duplicated?

Reply:
Fixed